Fix replays continuing after desync

PunkPun opened the pull request:
Overlooked in #19958

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ReplayControlBarLogic.cs:
```diff
@@ -46,17 +46,17 @@ public ReplayControlBarLogic(Widget widget, World world, OrderManager orderManag
+				playButton.IsVisible = () => world.ReplayTimestep == 0 || world.PauseStateLocked || orderManager.NetFrameNumber >= replayNetTicks;
 				playButton.OnClick = () => world.ReplayTimestep = (int)Math.Ceiling(originalTimestep * multipliers[speed]);
-				playButton.IsDisabled = () => orderManager.NetFrameNumber >= replayNetTicks;
+				playButton.IsDisabled = () => world.PauseStateLocked || orderManager.NetFrameNumber >= replayNetTicks;
```
These changes don't look correct - The `replayNetTicks` check is supposed to handle this, so that should either be fixed or removed.

PunkPun replied:
It isn't supposed to, `replayNetTicks` is a cached variable set to last replay tick. Going out of sync has nothing to do with replay parsing or replay length

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ReplayControlBarLogic.cs:
```diff
@@ -46,17 +46,17 @@ public ReplayControlBarLogic(Widget widget, World world, OrderManager orderManag
 
 				var pauseButton = widget.Get<ButtonWidget>("BUTTON_PAUSE");
-				pauseButton.IsVisible = () => world.ReplayTimestep != 0 && orderManager.NetFrameNumber < replayNetTicks;
+				pauseButton.IsVisible = () => world.ReplayTimestep != 0 && !world.PauseStateLocked && orderManager.NetFrameNumber < replayNetTicks;
```
Fwiw, just because `PauseStateLocked` is set, the world doesn't need to be paused.

PunkPun replied:
That is true, I'll make it pause also after a fatal lua crash and on mission objective fail

pchote replied:
I think @abcdefg30 means that the `PauseStateLocked` check here and on `playButton` should actually be on `.Disabled` not `.IsVisible`?

abcdefg30 replied:
Yes. I think we only lock the pause state after pausing at the moment, but how should the user know that and we probably shouldn't rely on that assumption ourselves either.

PunkPun replied:
2 buttons overlap, that's why there is the IsVisible in the first place. One button becomes visible while the other is non-visible

PunkPun replied:
PauseStateLocked was removed, using IsGameOver now instead

PunkPun committed:
Move force pausing to EndGame method

PunkPun committed:
Fix replays continuing after desync

PunkPun commented:
Since EndGame was used in the exact same places as ReplayTimestep needed to be set to 0, I moved it to the method 
and since PauseStateLocked always correlated with IsGameEnded, I removed and replaced it

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ReplayControlBarLogic.cs:
```diff
@@ -46,17 +46,17 @@ public ReplayControlBarLogic(Widget widget, World world, OrderManager orderManag
+				playButton.IsVisible = () => world.ReplayTimestep == 0 || world.IsGameOver || orderManager.NetFrameNumber >= replayNetTicks;
 				playButton.OnClick = () => world.ReplayTimestep = (int)Math.Ceiling(originalTimestep * multipliers[speed]);
-				playButton.IsDisabled = () => orderManager.NetFrameNumber >= replayNetTicks;
+				playButton.IsDisabled = () => world.IsGameOver || orderManager.NetFrameNumber >= replayNetTicks;
```
Are these `orderManager.NetFrameNumber >= replayNetTicks` checks obsoleted by `IsGameOver` now? If not, would it make sense to?

PunkPun replied:
These seem now obsolete in regards to desync or a regular game end. But all ways of ending a game should be tested, such as client connection being dropped

PunkPun replied:
https://github.com/OpenRA/OpenRA/pull/20670#issuecomment-1434472690

PunkPun commented:
Okay, last refactor. Relying on IsGameEnded or PauseStateLocked was a bad idea, as replays can have chat after the game has ended. Instead I made outOfSync flag in OrderManager public, and made ReplayControlBar disable its buttons while it's true.

The first commit is no longer necessary, it just cleanup. I can drop it if need be.

penev92 commented:
This button doesn't get updated initially and says "Resume" when the menu pops open upon a desync. If you press it, then reopen the menu it now says "Return to map", as it should from the start:
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs#L438

penev92 commented on OpenRA.Game/World.cs:
```diff
@@ -615,8 +618,7 @@ public void Dispose()
 			EndGame();
-			SetPauseState(true);
-			PauseStateLocked = true;
+			ReplayTimestep = 0;
```
Please add a comment like
```suggestion
			// Make sure the replay is completely frozen. Only set this here to not disable post-game chat in other situations.
			ReplayTimestep = 0;
```

PunkPun replied:
Added a comment

penev92 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ReplayControlBarLogic.cs:
```diff
@@ -86,7 +86,7 @@ public ReplayControlBarLogic(Widget widget, World world, OrderManager orderManag
 				var maximumButton = widget.Get<ButtonWidget>("BUTTON_MAXIMUM");
 				maximumButton.IsHighlighted = () => speed == PlaybackSpeed.Maximum;
-				maximumButton.IsDisabled = () => orderManager.NetFrameNumber >= replayNetTicks;
+				maximumButton.IsDisabled = () => orderManager.IsOutOfSync || orderManager.NetFrameNumber >= replayNetTicks;
```
We have this exact statement 6 times. I suggest adding a local function and using that everywhere:
```cs
bool ShouldDisableCommandButton() => orderManager.IsOutOfSync || orderManager.NetFrameNumber >= replayNetTicks;
```

Actually the Pause button's `IsVisible` could use a `!ShouldDisableCommandButton()`, so that's 7.

penev92 reviewed (commented):
Other than that lgtm.

abcdefg30 commented on OpenRA.Game/World.cs:
```diff
@@ -68,15 +68,19 @@ public void SetPlayers(IEnumerable<Player> players, Player localPlayer)
 		{
 			if (!IsGameOver)
 			{
+				SetPauseState(true);
```
Is there a reason you put this at the start of the method instead of at the end (which matches the previous behaviour)?

PunkPun replied:
Yes, because I made it so you can't change PausedState when IsGameOver is true

abcdefg30 replied:
Ah, makes sense.

pchote reviewed (commented):
Not tested, but code changes look reasonable

Mailaender reviewed (approved):
Tested in game. Works as promised.

Mailaender commented:
93d10f658e8e5f37287c865870de81ff9b314cdb
e8f8f7c730b9e29327548e98bf58df9cc674de5e

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/3d5c28b3f771dfd051b10b16d7c0ad8ea1cc9e91...b3097fa0d33263e4aeafac56ecbb05d8a434ce0c)
